go-controller/pkg/node/gateway_localnet.go

```diff
@@ -116,32 +122,107 @@ func generateGatewayNATRules(ifname string, ip net.IP) []iptRule {
 	return rules
 }
 
-func localnetGatewayNAT(ipt util.IPTablesHelper, ifname string, ip net.IP) error {
-	rules := generateGatewayNATRules(ifname, ip)
-	return addIptRules(ipt, rules)
+func localnetGatewayNAT(localNetdata []*localnetData, ifname string) error {
+	for _, lnData := range localNetdata {
+		rules := generateGatewayNATRules(ifname, lnData.gatewayIP)
+		err := addIptRules(lnData.ipt, rules)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFactory, nodeAnnotator kube.Annotator) error {
-	// Create a localnet OVS bridge.
+func initLocalnetGateway(nodeName string, subnets []*net.IPNet, wf *factory.WatchFactory, nodeAnnotator kube.Annotator) error {
+
 	localnetBridgeName := "br-local"
-	_, stderr, err := util.RunOVSVsctl("--may-exist", "add-br",
-		localnetBridgeName)
+
+	// Evaluate input subnets
+	err := evaluateInputNetworks(subnets)
 	if err != nil {
-		return fmt.Errorf("Failed to create localnet bridge %s"+
-			", stderr:%s (%v)", localnetBridgeName, stderr, err)
+		return err
 	}
 
-	ifaceID, macAddress, err := bridgedGatewayNodeSetup(nodeName, localnetBridgeName, localnetBridgeName, true)
+	// Setup local brifge and ports
+	ifaceID, macAddress, link, err := createAndSetupLocalBridgePorts(localnetBridgeName, nodeName)
 	if err != nil {
-		return fmt.Errorf("failed to set up shared interface gateway: %v", err)
+		return err
 	}
-	_, err = util.LinkSetUp(localnetBridgeName)
+
+	// Create IpData structure required for all subseqhent operations
+	ipData, err := constructIPData(subnets)
 	if err != nil {
 		return err
 	}
 
+	// Set the link with the addresses from IP object
+
+	err = addLinkAddresses(link, ipData)
```

Comment:
I think a lot of these helper functions make the code _less_ clear, because now I need to go look somewhere else to figure out exactly what it's doing. I would say for now, please un-split-out all of the functions you split out and just leave `initLocalnetGateway` big and sprawling. There are cleanups that could be made here, but I don't think the refactoring you're doing is in the spirit of golang coding style.

Reply:
I think we need to talk about this one. In my understanding there is nothing is that is not idiomatic go and I believe (respectfully)  we are better off with functions with clear responsibilities especially considering what the main method would have become with the dual stack support.

Reply:
@gvbalaji I agree with @danwinship here. For example: this function that you have added

```
func linkNeighExists(link netlink.Link, gatewayIP net.IP, macAddress net.HardwareAddr) bool {
	exists, err := util.LinkNeighExists(link, gatewayIP, macAddress)
	if err != nil {
		klog.Errorf("Error in exists call for MAC binding for %s on %s: %v. Will try to add the binding.",
			gatewayIP, localnetGatewayNextHopPort, err)
	}
	return exists
}
```

wraps the function just for logging an error. Why can't one directly use util.LinkNeighExists() instead of this wrapper?

Reply:
I could have. I did it this way because we are consciously ignoring the error if anything happens there and logging a message saying we will still go ahead with add to give us one more chance at it. Hence I thought it's a small unit by itself. Now if one wants to just unit test this method for both cases we could ( However I just started unit testing on this file. Yet to figure out how to make UT widely applicable  considering the file uses its dependencies all over the place)

Reply:
I made some pointed changes in this regard and resolving this.